harness/determined/pytorch/deepspeed/_deepspeed_context.py

```diff
@@ -0,0 +1,352 @@
+        self._per_slot_batch_size = self.env.per_slot_batch_size
+        self._train_micro_batch_size_per_gpu = None  # type: Optional[int]
+        self._num_micro_batches_per_slot = None  # type: Optional[int]
+        self._use_pipeline_parallel = False
+        self._data_repro_checks_disabled = False
+        self._manual_grad_accumulation = False
+
+        self._check_experiment_config()
+
+    def _check_experiment_config(self) -> None:
+        """
+        Check if the user specified options in optimizations that are incompatible with
+        DeepSpeedTrial.
+        """
+        optimizations_config = self.env.experiment_config.get_optimizations_config()
+        if not optimizations_config.get("average_training_metrics", False):
+            logging.warning(
+                "DeepSpeedTrial always averages training metrics across data parallel ranks."
+            )
+        self._average_training_metrics = True
+        mixed_precision_val = optimizations_config.get("mixed_precision", "O0")
+        if mixed_precision_val != "O0":
+            raise det.errors.InvalidExperimentException(
+                "Mixed precision is specified through the deepspeed config instead of the "
+                "Determined experiment config.",
+            )
+        aggregation_frequency = optimizations_config.get("aggregation_frequency", 1)
+        if aggregation_frequency > 1:
+            raise det.errors.InvalidExperimentException(
+                "Gradient aggregation is specified through the deepspeed config instead of the "
+                "Determined experiment config.",
+            )
+
+    def get_global_batch_size(self) -> int:
+        return self._global_batch_size
+
+    def get_per_slot_batch_size(self) -> int:
+        return self._per_slot_batch_size
+
+    def wrap_mpu(self, mpu: det_ds.ModelParallelUnit) -> det_ds.ModelParallelUnit:
+        if self._called_wrap_mpu:
+            raise det.errors.InvalidExperimentException(
+                "Only one MPU can be passed to DeepSpeedTrialContext.  "
+                "Please make sure wrap_mpu is only called once in the trial definition."
+            )
+        self._called_wrap_mpu = True
+        old_mpu = self._mpu
+        self._mpu = mpu
+        if old_mpu.get_data_parallel_world_size() != mpu.get_data_parallel_world_size():
+            (
+                self._per_slot_batch_size,
+                self._global_batch_size,
+            ) = self._calculate_batch_sizes()
+            logging.warning(
+                "Wrapped MPU uses model parallelism.  Changing per slot batch size "
+                f"to {self.get_per_slot_batch_size()} and "
+                f"global batch size to {self.get_global_batch_size()}"
+            )
+        return mpu
+
+    @property
+    def mpu(self) -> det_ds.ModelParallelUnit:
+        return self._mpu
+
+    def wrap_model_engine(self, model: torch.nn.Module) -> torch.nn.Module:
```

Comment:
can `wrap_model_engine()` be called twice by the user?  What happens if the values we extract from the model don't match?

Also, shouldn't this take/return a different type than `nn.Module`?

Reply:
Currently, the plan is to support multiple model engines for training without model parallelism or training with the same mpu.  The record keeping for num_inputs will be based on the first model engine that is passed in and if gradient_aggregation steps differ between model engines, the user will have to set context.disable_auto_grad_accumulation.  This will all be included in the docs.